engine/crates/operation-checks/src/schema/wrapper_types.rs

```diff
@@ -0,0 +1,104 @@
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+#[repr(u8)]
+pub enum WrapperType {
+    List = 0b10,
+    Required = 0b01,
+    RequiredList = 0b11,
+}
+
+impl WrapperType {
+    const LIST: u8 = WrapperType::List as u8;
+    const REQUIRED: u8 = WrapperType::Required as u8;
+    const REQUIRED_LIST: u8 = WrapperType::RequiredList as u8;
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum WrapperTypes {
+    /// A compact representation for up to four levels of wrapper types. Each pair of bits
+    /// corresponds to an Option<WrapperType> (0b00 = None, 0b01 = Some(Required), 0b10 =>
+    /// Some(List), 0b11 => Some(RequiredList)). The least significant pair of bits corresponds to
+    /// the outermost wrappers.
+    Small(u8),
+    /// This variant is for extreme cases of list fields with more than three levels of nested
+    /// lists. From outermost to innermost.
+    Large(Box<[WrapperType]>),
+}
+
+impl WrapperTypes {
+    /// Iterate wrapper types from outermost to innermost.
+    pub(crate) fn iter_wrappers(&self) -> impl Iterator<Item = WrapperType> + '_ {
+        let (first_four, rest): ([Option<WrapperType>; 4], &[WrapperType]) = match self {
+            WrapperTypes::Small(small) => (
+                [0, 1, 2, 3].map(move |i| {
+                    let shift = i * 2;
+                    let bits = (small >> shift) & 0b11;
+
+                    match bits {
+                        0b00 => None,
+                        WrapperType::REQUIRED => Some(WrapperType::Required),
+                        WrapperType::LIST => Some(WrapperType::List),
+                        WrapperType::REQUIRED_LIST => Some(WrapperType::RequiredList),
+                        _ => unreachable!(),
+                    }
+                }),
+                &[],
+            ),
+            WrapperTypes::Large(large) => ([None; 4], large.as_ref()),
+        };
+
+        first_four.into_iter().flatten().chain(rest.iter().copied())
```

Comment:
The dance with the array is here to avoid having to box the iterator. Open to better ideas.